File: manager/src/main/java/io/hydrosphere/serving/manager/util/IsolatedDockerClient.java
package io.hydrosphere.serving.manager.util;


import com.spotify.docker.client.DefaultDockerClient;
import com.spotify.docker.client.ProgressHandler;
import com.spotify.docker.client.exceptions.DockerCertificateException;
import com.spotify.docker.client.exceptions.DockerException;
import com.spotify.docker.client.messages.ContainerConfig;
import com.spotify.docker.client.messages.ContainerCreation;

import java.io.IOException;
import java.nio.file.Path;
import java.util.concurrent.ConcurrentSkipListSet;

public class IsolatedDockerClient extends DefaultDockerClient {

    public static IsolatedDockerClient createFromEnv() throws DockerCertificateException {
        return new IsolatedDockerClient(DefaultDockerClient.fromEnv());
    }

    private ConcurrentSkipListSet<String> imageStorage = new ConcurrentSkipListSet<>();
    private ConcurrentSkipListSet<String> containerStorage = new ConcurrentSkipListSet<>();

    private IsolatedDockerClient(Builder builder) {
        super(builder);
    }

    @Override
    public String build(Path directory, String name, String dockerfile, ProgressHandler handler, BuildParam... params)
            throws DockerException, InterruptedException, IOException {
        String image = super.build(directory, name, dockerfile, handler, params);
        imageStorage.add(image);
        return image;
    }

    @Override
    public ContainerCreation createContainer(final ContainerConfig config, final String name)
            throws DockerException, InterruptedException {
        ContainerCreation creation = super.createContainer(config, name);
        containerStorage.add(creation.id());
        return creation;
    }

    public void clear() {
        for (String container : containerStorage) {
            try {
                super.removeContainer(container, RemoveContainerParam.forceKill());
                containerStorage.remove(container);
            } catch (DockerException | InterruptedException e) {
                e.printStackTrace();
            }
        }
        for (String image : imageStorage) {
            try {
                super.removeImage(image, true, true);
                imageStorage.remove(image);
            } catch (DockerException | InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void close() {
        clear();
        super.close();
    }
}

File: manager/src/main/scala/io/hydrosphere/serving/manager/util/docker/IsolatedDockerClient.scala
package io.hydrosphere.serving.manager.util.docker


import com.spotify.docker.client.{DefaultDockerClient, DockerClient, ProgressHandler}
import com.spotify.docker.client.exceptions.DockerCertificateException
import com.spotify.docker.client.exceptions.DockerException
import com.spotify.docker.client.messages.ContainerConfig
import com.spotify.docker.client.messages.ContainerCreation
import java.io.IOException
import java.nio.file.Path
import java.util.concurrent.ConcurrentSkipListSet

import scala.collection.JavaConverters._
import com.spotify.docker.client.DockerClient.RemoveContainerParam


object IsolatedDockerClient {
  def createFromEnv = new IsolatedDockerClient(DefaultDockerClient.fromEnv)
}

class IsolatedDockerClient private(val builder: DefaultDockerClient.Builder) extends DefaultDockerClient(builder) {
  private val imageStorage = new ConcurrentSkipListSet[String]
  private val containerStorage = new ConcurrentSkipListSet[String]

  override def build(directory: Path, name: String, dockerfile: String, handler: ProgressHandler, params: DockerClient.BuildParam*): String = {
    val image = super.build(directory, name, dockerfile, handler, params:_*)
    imageStorage.add(image)
    image
  }

  override def removeContainer(containerId: String, params: RemoveContainerParam*): Unit = {
    super.removeContainer(containerId, params: _*)
    containerStorage.remove(containerId)
  }

  override def createContainer(config: ContainerConfig, name: String): ContainerCreation = {
    val creation = super.createContainer(config, name)
    containerStorage.add(creation.id)
    creation
  }

  def clear(): Unit = {
    for (container <- containerStorage.asScala) {
      try {
        super.removeContainer(container, RemoveContainerParam.forceKill)
        containerStorage.remove(container)
      } catch {
        case e@(_: DockerException | _: InterruptedException) =>
          e.printStackTrace()
      }
    }
    for (image <- imageStorage.asScala) {
      try {
        super.removeImage(image, true, true)
        imageStorage.remove(image)
      } catch {
        case e@(_: DockerException | _: InterruptedException) =>
          e.printStackTrace()
      }
    }
  }

  override def close(): Unit = {
    clear()
    super.close()
  }
}
